File: typescript/docs/examples/browser-use/browser/ws_push_callback_captcha_tongcheng.ts
/**
 * Example demonstrating how to receive backend push notifications via Session WS.
 *
 * This example shows how to:
 * - Create a browser session
 * - Connect to the session WS endpoint
 * - Register a push callback for target: wuying_cdp_mcp_server
 * - Trigger a captcha flow on Tongcheng and wait for backend push
 */

import { AgentBay } from "wuying-agentbay-sdk";
import * as playwright from "playwright";

async function main(): Promise<void> {
  const apiKey = process.env.AGENTBAY_API_KEY;
  if (!apiKey) {
    throw new Error("AGENTBAY_API_KEY environment variable not set");
  }

  const imageId = process.env.AGENTBAY_WS_IMAGE_ID || "imgc-0ab5ta4kuo0x3pa70";
  const agentBay = new AgentBay({ apiKey });
  const created = await agentBay.create({ imageId });
  if (!created.success || !created.session) {
    throw new Error(`Failed to create session: ${created.errorMessage}`);
  }
  const session = created.session;

  try {
    const wsClient = await session.getWsClient();
    const pushPromise = new Promise<void>((resolve, reject) => {
      const t = setTimeout(() => reject(new Error("timeout waiting for WS push")), 180000);
      wsClient.registerCallback("wuying_cdp_mcp_server", (payload) => {
        clearTimeout(t);
        process.stdout.write(`WS PUSH: ${JSON.stringify(payload)}\n`);
        resolve();
      });
    });

    const ok = await session.browser.initializeAsync({ useStealth: true, solveCaptchas: true });
    if (!ok) {
      throw new Error("Failed to initialize browser");
    }
    const endpointUrl = await session.browser.getEndpointUrl();

    const browser = await playwright.chromium.connectOverCDP(endpointUrl);
    try {
      const context = browser.contexts()[0];
      const page = await context.newPage();
      await page.goto("https://passport.ly.com/Passport/GetPassword", { waitUntil: "domcontentloaded" });
      const input = await page.waitForSelector("#name_in", { timeout: 10000 });
      await input.click();
      await input.fill("");
      await input.type("13000000000");
      await page.waitForTimeout(1000);
      await page.click("#next_step1");

      process.stdout.write("Waiting for backend push...\n");
      await pushPromise;
      process.stdout.write("Received backend push.\n");
    } finally {
      await browser.close();
    }
  } finally {
    await agentBay.delete(session);
  }
}

if (require.main === module) {
  main().catch((e) => {
    process.stderr.write(String(e) + "\n");
    process.exitCode = 1;
  });
}

File: python/docs/examples/_async/browser-use/browser/ws_push_callback_captcha_tongcheng.py
"""
Example demonstrating how to receive backend push notifications via Session WS.

This example shows how to:
- Create a browser session
- Connect to the session WS endpoint
- Register a push callback for target: wuying_cdp_mcp_server
- Trigger a captcha flow on Tongcheng and wait for backend push
"""

import asyncio
import os

from playwright.async_api import async_playwright

from agentbay import AsyncAgentBay, BrowserOption, CreateSessionParams


async def main() -> None:
    api_key = os.getenv("AGENTBAY_API_KEY")
    if not api_key:
        raise RuntimeError("AGENTBAY_API_KEY environment variable not set")

    image_id = os.getenv("AGENTBAY_WS_IMAGE_ID") or "imgc-0ab5ta4kuo0x3pa70"
    agent_bay = AsyncAgentBay(api_key=api_key)
    created = await agent_bay.create(CreateSessionParams(image_id=image_id))
    if not created.success or created.session is None:
        raise RuntimeError(f"Failed to create session: {created.error_message}")
    session = created.session

    ws_client = None
    try:
        ws_client = await session._get_ws_client()
        push_signal = asyncio.Event()

        def on_push(payload: dict) -> None:
            print(f"WS PUSH: {payload}")
            push_signal.set()

        ws_client.register_callback("wuying_cdp_mcp_server", on_push)
        await ws_client.connect()

        browser_option = BrowserOption(use_stealth=True, solve_captchas=True)
        ok = await session.browser.initialize(browser_option)
        if not ok:
            raise RuntimeError("Failed to initialize browser")

        endpoint_url = await session.browser.get_endpoint_url()
        if not endpoint_url:
            raise RuntimeError("Empty browser endpoint URL")

        async with async_playwright() as p:
            browser = await p.chromium.connect_over_cdp(endpoint_url)
            page = await browser.contexts[0].new_page()
            await page.goto(
                "https://passport.ly.com/Passport/GetPassword",
                wait_until="domcontentloaded",
            )
            input_element = await page.wait_for_selector("#name_in", timeout=10000)
            await input_element.click()
            await input_element.fill("")
            await input_element.type("13000000000")
            await page.wait_for_timeout(1000)
            await page.click("#next_step1")

            print("Waiting for backend push...")
            await asyncio.wait_for(push_signal.wait(), timeout=180.0)
            print("Received backend push.")
    finally:
        if ws_client is not None:
            try:
                await ws_client.close()
            except Exception:
                pass
        await session.delete()


if __name__ == "__main__":
    asyncio.run(main())

